mlrun/db/filedb.py

```diff
@@ -180,19 +184,59 @@ def del_artifacts(self, name='', project='', tag='', labels=[]):
 
                 self._safe_del(p)
 
+    def store_function(self, func, name, project='', tag=''):
+        data = self._dumps(func)
+        filepath = self._filepath(functions_dir, project, name) + self.format
+        self._datastore.put(filepath, data)
+        filepath = self._filepath(
+            functions_dir, project, name, tag or 'latest') + self.format
+        self._datastore.put(filepath, data)
+
+    def get_function(self, name, project='', tag=''):
+        filepath = self._filepath(
+            functions_dir, project, name, tag) + self.format
+        data = self._datastore.get(filepath)
+        return self._loads(data)
+
+    def list_functions(self, name, project='', tag='', labels=None):
+        tag = tag or 'latest'
+        labels = labels or []
+        logger.info(
+            f'reading functions in {project} name/mask: {name} tag: {tag} ...')
+        filepath = self._filepath(functions_dir, project, tag=tag)
+        results = FunctionList(tag)
+        if isinstance(labels, str):
+            labels = labels.split(',')
+        if tag == '*':
+            mask = '**/*' + name
+            if name:
+                mask += '*'
+        else:
+            mask = '**/*'
+        for func, p in self._load_list(filepath, mask):
+            if (name == '' or name in get_in(func, 'name', ''))\
+                    and match_labels(get_in(func, 'labels', {}), labels):
+                if 'artifacts/latest' in p:
+                    func['tree'] = 'latest'
+                results.append(func)
+
+        return results
+
     def _filepath(self, table, project, key='', tag=''):
         if tag == '*':
             tag = ''
         if tag:
             key = '/' + key
         if project:
-            return path.join(self.dirpath, '{}/{}/{}{}'.format(table, project, tag, key))
+            return path.join(self.dirpath, '{}/{}/{}{}'.format(
+                table, project, tag, key))
         else:
             return path.join(self.dirpath, '{}/{}{}'.format(table, tag, key))
 
     def _dumps(self, obj):
         if self.format == '.yaml':
```

Comment:
all the objects are passed to DB as dicts, so need to use `dict_to_yaml(struct)` and `dict_to_json(struct)`

Reply:
Don't you want to support an object (such as an `Artifact`)? IMO it's more natural - we can do it for the user.

Reply:
@tebeka it was like that, but wont work in case of forwarding, there can be many sub classes, so better keep just the dict and pass along